### PR TITLE
FIX some bugs in dex.c, marshal.c, panels.c and other code optimization and fixel

### PR DESCRIPTION
**Pull request by ChD1**
**Checklist**

- [ ] Closing issues: #issue
- [ ] Mark this if you consider it ready to merge
- [ ] I've added tests (optional)
- [ ] I wrote some lines in the [radare2book](https://github.com/radareorg/radare2book)

**Description**

### CODE REVIEW

File: libr/main/rabin2.c [hunk 1]
```diff
@@ -887,7 +887,10 @@ R_API int r_main_rabin2(int argc, const char **argv) {
 				R_FREE (res);
 				R_FREE (file);
 			}
-			stdin_gets (true);
+			char *r = stdin_gets (true);
+			if(r){
```
Comment (trufae) [inline libr/main/rabin2.c:891]:
```suggestion
			if (r) {
```

File: libr/core/panels.c [hunk 2]
```diff
@@ -1067,6 +1067,9 @@ static void __set_rcb(RPanels *ps, RPanel *p) {
 }
 
 static void __init_panel_param(RCore *core, RPanel *p, const char *title, const char *cmd) {
+	if(!p){
```
Comment (trufae) [inline libr/core/panels.c:1070]:
```suggestion
	if (!p) {
```

Comment (trufae) [inline libr/core/panels.c]:
```suggestion
	if (!p) {
```

File: libr/core/panels.c [hunk 3]
```diff
@@ -1554,22 +1557,34 @@ static void __cursor_down(RCore *core) {
 }
 
 static void __save_panel_pos(RPanel* panel) {
+	if(!panel){
```
Comment (trufae) [inline libr/core/panels.c:1560]:
```suggestion
	if (!panel) {
```

File: libr/core/panels.c [hunk 4]
```diff
@@ -1554,22 +1557,34 @@ static void __cursor_down(RCore *core) {
...
 	RPanel *cur = __get_cur_panel (panels);
+	if(!cur){
+		return;
+	}
 	__set_geometry (&cur->view->pos, 0, 1, panels->can->w, panels->can->h - 1);
 	cur->view->refresh = true;
 }
 
 static void __dismantle_panel(RPanels *ps, RPanel *p) {
+	if(!p){
```
Comment (trufae) [inline libr/core/panels.c:1585]:
```suggestion
	if (!p) {
```

File: libr/core/panels.c [hunk 5]
```diff
@@ -1827,6 +1842,9 @@ static void __init_sdb(RCore *core) {
 }
 
 static void __free_panel_model(RPanel *panel) {
+	if(!panel){
```
Comment (trufae) [inline libr/core/panels.c:1845]:
```suggestion
	if (!panel) {
```

File: libr/core/panels.c [hunk 6]
```diff
@@ -1855,6 +1873,9 @@ static void __create_panel(RCore *core, RPanel *panel, const RPanelLayout dir, R
 	if (!__check_panel_num (core)) {
 		return;
 	}
+	if(!panel){
```
Comment (trufae) [inline libr/core/panels.c:1876]:
```suggestion
	if (!panel) {
```

File: libr/core/vmenus.c [hunk 7]
```diff
@@ -785,11 +785,19 @@ R_API int r_core_visual_types(RCore *core) {
...
 			} else {
 				switch (h_opt) {
-				case 0: // type
+				case 0:{ // type
 					/* TODO: do something with this data */
-					prompt ("name: ", vt.curname);
-					prompt ("pf: ", vt.curfmt);
+					char *r = NULL;
+					r = prompt ("name: ", vt.curname);
+					if(r){
```
Comment (trufae) [inline libr/core/vmenus.c:792]: just call free (r); the if is unnecessary

File: libr/core/vmenus.c [hunk 8]
```diff
@@ -785,11 +785,19 @@ R_API int r_core_visual_types(RCore *core) {
...
+						free(r);
+					}
+					r = prompt ("pf: ", vt.curfmt);
+					if(r){
+						free(r);
```
Comment (trufae) [inline libr/core/vmenus.c:797]: same here

File: libr/main/rabin2.c [hunk 9]
```diff
@@ -887,7 +887,10 @@ R_API int r_main_rabin2(int argc, const char **argv) {
 				R_FREE (res);
 				R_FREE (file);
 			}
-			stdin_gets (true);
+			char *r = stdin_gets (true);
+			if (r) {
```
Comment (trufae) [inline libr/main/rabin2.c:891]: same here

**Review by trufae — CHANGES_REQUESTED** [review]
Please git pull the commits i pushed before addressing the comments. free(NULL) is valid, so there's no need to wrap it with an if()

File: libr/main/radare2.c [hunk 10]
```diff
@@ -587,6 +589,7 @@ R_API int r_main_radare2(int argc, const char **argv) {
 				r_core_project_list (r, 0);
 				r_cons_flush ();
 				free (envprofile);
+                free (debugbackend);
```
Comment (trufae) [inline libr/main/radare2.c:592]: bad indentation. use tabs

File: libr/main/radare2.c [hunk 11]
```diff
@@ -532,6 +533,7 @@ R_API int r_main_radare2(int argc, const char **argv) {
 		case 'H':
 			main_print_var (opt.arg);
 			LISTS_FREE ();
+            free (debugbackend);
```
Comment (trufae) [inline libr/main/radare2.c:536]: use tabs

File: libr/main/rahash2.c [hunk 12]
```diff
@@ -653,6 +653,10 @@ R_API int r_main_rahash2(int argc, const char **argv) {
 			if (rad == 'j') {
 				pj = pj_new ();
 				if (!pj) {
+                    if (str != hashstr) {
```
Comment (trufae) [inline libr/main/rahash2.c:656]: use tabs

File: libr/util/pkcs7.c [hunk 13]
```diff
@@ -326,6 +326,7 @@ R_API RCMS *r_pkcs7_parse_cms(const ut8 *buffer, ut32 length) {
 		container->contentType = r_asn1_stringify_oid (object->list.objects[0]->sector, object->list.objects[0]->length);
 		if (!container->contentType) {
 			r_asn1_free_object (object);
+            free (container);
```
Comment (trufae) [inline libr/util/pkcs7.c:329]: use tabs

File: libr/core/linux_heap_glibc.c [hunk 14]
```diff
@@ -1122,6 +1123,11 @@ static void GH(print_heap_segment)(RCore *core, MallocState *main_arena,
 
 	if (!r_io_read_at (core->io, next_chunk, (ut8 *)cnk, sizeof (GH(RHeapChunk)))) {
 		eprintf ("Cannot read");
+        free (cnk);
```
Comment (trufae) [inline libr/core/linux_heap_glibc.c:1126]: use tabs

File: libr/cons/pal.c [hunk 15]
```diff
@@ -617,7 +617,10 @@ R_API int r_cons_pal_set(const char *key, const char *val) {
 	for (i = 0; keys[i].name; i++) {
 		if (!strcmp (key, keys[i].name)) {
 			rcolor = RCOLOR_AT (i);
-			r_cons_pal_parse (val, rcolor);
+			char *r = r_cons_pal_parse (val, rcolor);
+            if (r) {
```
Comment (trufae) [inline libr/cons/pal.c:621]: unnecessary if

File: libr/core/anal_tp.c [hunk 16]
```diff
@@ -507,10 +507,12 @@ R_API void r_core_anal_type_match(RCore *core, RAnalFunction *fcn) {
 	char *ret_reg = NULL;
 	const char *pc = r_reg_get_name (core->dbg->reg, R_REG_NAME_PC);
 	if (!pc) {
+        free (buf);
 		return;
 	}
 	RRegItem *r = r_reg_get (core->dbg->reg, pc, -1);
 	if (!r) {
+        free (buf);
```
Comment (trufae) [inline libr/core/anal_tp.c:515]: use tabs

File: libr/core/cmd_debug.c [hunk 17]
```diff
@@ -2681,6 +2681,7 @@ static void cmd_debug_reg(RCore *core, const char *str) {
 					break;
 				default:
 					eprintf ("Unkown comamnd");
+                    free (name);
```
Comment (trufae) [inline libr/core/cmd_debug.c:2684]: tabs

File: libr/core/cmd_open.c [hunk 18]
```diff
@@ -1926,7 +1926,9 @@ static int cmd_open(void *data, const char *input) {
 			if ((fdx == -1) || (fd == -1) || (fdx == fd)) {
 				free (inp);
 				break;
-			}
+			} else {
```
Comment (trufae) [inline libr/core/cmd_open.c:1929]: tabs

**Review by trufae — CHANGES_REQUESTED** [review]
Please use tabs

File: libr/core/cmd_open.c (hunk 18 above)
Comment (ChD1) [reply libr/core/cmd_open.c:1929]: All done.

**Comment by trufae** [comment]
Fix the warning to make the ci happy
470DAE1F-192C-4090-9A6B-F6BAC08D803A

**Comment by ChD1** [comment]
Done. I think.

**Comment by trufae** [comment]
Perfect!! Good findings!

**Pull request merged**